[Frontend][Parse] Parse new named opaque return type syntax

willtunnels opened the pull request:
First commit in the process of implementing the proposal for named opaque return types and associated 'where' clause support. See discussion here: https://forums.swift.org/t/improving-the-ui-of-generics/22814.

Commit:
[Frontend][Parse] Parse new named opaque return type syntax

In order to put constraints on opaque types in function returns, we want to
support naming them like 'func f() -> <T> T { }'. This commit parses that
syntax into the new `OpaqueReturnParameteriedTypeRepr`. This is hidden behind
the new flag --enable-experimental-opaque-return-types.

willtunnels commented:
@swift-ci please test

slavapestov commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -864,17 +865,39 @@ Parser::parseFunctionSignature(Identifier SimpleName,
       return Status;
 
+    if (GenericParams != nullptr) {
+      // The `Base` for our `OpaqueReturnParameterizedTypeRepr` should not be
```
The assert should go in the OpaqueReturnParameterizedTypeRepr constructor

slavapestov commented on test/type/opaque_experimental.swift:
```diff
@@ -0,0 +1,5 @@
+// RUN: %target-swift-frontend -enable-experimental-opaque-return-types -disable-availability-checking -typecheck -verify %s
```
Nitpick: you can use %target-typecheck-verify-swift to expand to %target-swift-frontend -typecheck -verify %s

slavapestov commented on test/type/opaque_experimental.swift:
```diff
@@ -0,0 +1,5 @@
+
+// Tests for experimental extensions to opaque return type support.
+
+func f0() -> <T> () { }
```
Can you add tests with multiple generic parameters, and an inheritance clause on a generic parameter, as well as some invalid cases.

slavapestov commented on include/swift/AST/TypeRepr.h:
```diff
@@ -1101,6 +1101,42 @@ class OpaqueReturnTypeRepr : public TypeRepr {
+/// constraints that the concrete types satisfy:
+///
+/// func foo() -> <T: Collection> T { return [1] }
+class OpaqueReturnParameterizedTypeRepr : public TypeRepr {
```
Nitpick: Maybe NamedOpaqueReturnTypeRepr? "Parametrized" is rather long

slavapestov reviewed (commented):
Looks like a great start!

Commit:
[IDE][Parse] Address CR for new named opaque return type syntax

slavapestov commented on test/IDE/print_ast_named_opaque_return.swift:
```diff
@@ -0,0 +1,10 @@
+func f() -> <T> () {
+}
+func g() -> <T> Int {
```
Should try printing a generic parameter list with an inheritance clause as well

slavapestov commented on include/swift/AST/TypeRepr.h:
```diff
@@ -1101,6 +1101,44 @@ class OpaqueReturnTypeRepr : public TypeRepr {
+/// A TypeRepr for a type with a generic parameter list of named opaque return
+/// types.
+///
+/// This can occur only as the return type of a function declaration, to specify
```
It can also occur as the return type of a computed property or subscript, no? (Those should be tested as well)

Commit:
[Parse] Allow named opaque types in more places

- Allow named opaque types in typed patterns and subscripts
- Fix inheritance clause printing for `GenericParamList`
- clang-format changes from previous commit on this branch

slavapestov reviewed (approved):
Looks ready to merge, except for the var/subscript decl tests I mentioned.

Commit:
Add named opaque type test cases for patterns and subscripts

willtunnels commented:
@swift-ci please smoke test
